File: algorithms/sorting/Solutions.java
import java.util.*;

/** Sorting Algorithms — Essential Problem Solutions. */
public class Solutions {

    /** Stable divide-and-conquer sort. O(n log n) time, O(n) space. */
    public static int[] mergeSort(int[] arr) {
        if (arr.length <= 1) return arr;
        int mid = arr.length / 2;
        int[] left = mergeSort(Arrays.copyOfRange(arr, 0, mid));
        int[] right = mergeSort(Arrays.copyOfRange(arr, mid, arr.length));
        return merge(left, right);
    }

    private static int[] merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int i = 0, j = 0, k = 0;
        while (i < left.length && j < right.length) {
            result[k++] = (left[i] <= right[j]) ? left[i++] : right[j++];
        }
        while (i < left.length) result[k++] = left[i++];
        while (j < right.length) result[k++] = right[j++];
        return result;
    }

    /** Merge Intervals — sort by start, single-pass merge. O(n log n). */
    public static int[][] mergeIntervals(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        List<int[]> merged = new ArrayList<>();
        for (int[] interval : intervals) {
            if (!merged.isEmpty() && interval[0] <= merged.get(merged.size() - 1)[1]) {
                merged.get(merged.size() - 1)[1] =
                        Math.max(merged.get(merged.size() - 1)[1], interval[1]);
            } else {
                merged.add(interval);
            }
        }
        return merged.toArray(new int[0][]);
    }

    /** Kth Largest Element via Quickselect. O(n) average. */
    public static int findKthLargest(int[] nums, int k) {
        int target = nums.length - k;
        int lo = 0, hi = nums.length - 1;
        Random rand = new Random();
        while (lo <= hi) {
            int p = partition(nums, lo, hi, rand);
            if (p == target) return nums[p];
            if (p < target) lo = p + 1;
            else hi = p - 1;
        }
        return nums[lo];
    }

    private static int partition(int[] nums, int lo, int hi, Random rand) {
        int pivotIdx = lo + rand.nextInt(hi - lo + 1);
        swap(nums, pivotIdx, hi);
        int pivot = nums[hi], store = lo;
        for (int i = lo; i < hi; i++) {
            if (nums[i] <= pivot) swap(nums, store++, i);
        }
        swap(nums, store, hi);
        return store;
    }

    private static void swap(int[] nums, int i, int j) {
        int tmp = nums[i]; nums[i] = nums[j]; nums[j] = tmp;
    }

    /** Dutch National Flag — three-way in-place partition. O(n) time. */
    public static void sortColors(int[] nums) {
        int low = 0, mid = 0, high = nums.length - 1;
        while (mid <= high) {
            if (nums[mid] == 0) {
                swap(nums, low++, mid++);
            } else if (nums[mid] == 1) {
                mid++;
            } else {
                swap(nums, mid, high--);
            }
        }
    }

    /** Meeting Rooms II — sort starts/ends, two-pointer sweep. O(n log n). */
    public static int minMeetingRooms(int[][] intervals) {
        if (intervals.length == 0) return 0;
        int[] starts = new int[intervals.length];
        int[] ends = new int[intervals.length];
        for (int i = 0; i < intervals.length; i++) {
            starts[i] = intervals[i][0];
            ends[i] = intervals[i][1];
        }
        Arrays.sort(starts);
        Arrays.sort(ends);
        int rooms = 0, maxRooms = 0, s = 0, e = 0;
        while (s < starts.length) {
            if (starts[s] < ends[e]) {
                rooms++;
                maxRooms = Math.max(maxRooms, rooms);
                s++;
            } else {
                rooms--;
                e++;
            }
        }
        return maxRooms;
    }

    /** Custom Sort String — frequency map + bucket ordering. O(n + k). */
    public static String customSortString(String order, String s) {
        int[] count = new int[26];
        for (char ch : s.toCharArray()) count[ch - 'a']++;
        StringBuilder result = new StringBuilder();
        for (char ch : order.toCharArray()) {
            result.append(String.valueOf(ch).repeat(count[ch - 'a']));
            count[ch - 'a'] = 0;
        }
        for (int i = 0; i < 26; i++) {
            result.append(String.valueOf((char) ('a' + i)).repeat(count[i]));
        }
        return result.toString();
    }
}

File: algorithms/sorting/solutions.py
"""Sorting Algorithms — Essential Problem Solutions."""

from typing import List


def merge_sort(arr: List[int]) -> List[int]:
    """Stable divide-and-conquer sort. O(n log n) time, O(n) space."""
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return _merge(left, right)


def _merge(left: List[int], right: List[int]) -> List[int]:
    result: List[int] = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


def quick_sort(arr: List[int]) -> List[int]:
    """In-place quicksort with random pivot. O(n log n) average."""
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    mid = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + mid + quick_sort(right)


def merge_intervals(intervals: List[List[int]]) -> List[List[int]]:
    """Merge Intervals — sort by start, single-pass merge. O(n log n)."""
    intervals.sort(key=lambda x: x[0])
    merged: List[List[int]] = []
    for start, end in intervals:
        if merged and start <= merged[-1][1]:
            merged[-1][1] = max(merged[-1][1], end)
        else:
            merged.append([start, end])
    return merged


def find_kth_largest(nums: List[int], k: int) -> int:
    """Kth Largest Element via Quickselect. O(n) average."""
    import random

    def partition(lo: int, hi: int) -> int:
        pivot_idx = random.randint(lo, hi)
        nums[pivot_idx], nums[hi] = nums[hi], nums[pivot_idx]
        pivot = nums[hi]
        store = lo
        for i in range(lo, hi):
            if nums[i] <= pivot:
                nums[store], nums[i] = nums[i], nums[store]
                store += 1
        nums[store], nums[hi] = nums[hi], nums[store]
        return store

    target = len(nums) - k
    lo, hi = 0, len(nums) - 1
    while lo <= hi:
        p = partition(lo, hi)
        if p == target:
            return nums[p]
        if p < target:
            lo = p + 1
        else:
            hi = p - 1
    return nums[lo]


def sort_colors(nums: List[int]) -> None:
    """Dutch National Flag — three-way in-place partition. O(n) time."""
    low = mid = 0
    high = len(nums) - 1
    while mid <= high:
        if nums[mid] == 0:
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1
        elif nums[mid] == 1:
            mid += 1
        else:
            nums[mid], nums[high] = nums[high], nums[mid]
            high -= 1


def min_meeting_rooms(intervals: List[List[int]]) -> int:
    """Meeting Rooms II — sort starts/ends, two-pointer sweep. O(n log n)."""
    if not intervals:
        return 0
    starts = sorted(i[0] for i in intervals)
    ends = sorted(i[1] for i in intervals)
    rooms = max_rooms = 0
    s = e = 0
    while s < len(starts):
        if starts[s] < ends[e]:
            rooms += 1
            max_rooms = max(max_rooms, rooms)
            s += 1
        else:
            rooms -= 1
            e += 1
    return max_rooms


def custom_sort_string(order: str, s: str) -> str:
    """Custom Sort String — frequency map + bucket ordering. O(n + k)."""
    from collections import Counter

    count = Counter(s)
    result: List[str] = []
    for ch in order:
        result.append(ch * count.pop(ch, 0))
    for ch, freq in count.items():
        result.append(ch * freq)
    return "".join(result)
